CDAP-13147 documentation on operations dashboard and reporting

gsps1 opened the pull request:
Docs quick build : https://builds.cask.co/browse/CDAP-DQB419-9

Quick Links to documentation pages from build 

Overview page : https://builds.cask.co/artifact/CDAP-DQB419/shared/build-9/Docs-HTML/html/admin-manual/operations/operations-dashboard.html

Reporting REST API : https://builds.cask.co/artifact/CDAP-DQB419/shared/build-9/Docs-HTML/html/reference-manual/http-restful-api/reporting.html

Dashboard REST API page : https://builds.cask.co/artifact/CDAP-DQB419/shared/build-9/Docs-HTML/html/reference-manual/http-restful-api/operations.html

TODO : Will add section on limitation (reporting only on spark2, etc)
Edit : updated links to latest build

gsps1 committed:
CDAP-13147 documentation on opertaions dashboard and reports

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Operations dashboard provides users two views: Dashboard and Reporting.
+
+The Dashboard view is a real-time interactive interface that allows users to take a sneak peek
+of the program runs statistics, resource usage within any 24 hours in the last week.
```
I think this sounds better:
"any 24 hours" --> "any 24 hour time window"

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+of the program runs statistics, resource usage within any 24 hours in the last week.
+A user can also select any hour within the 24 hours view to get program run details within that hour.
+
+The Reporting view provides a comprehensive report of program runs, users can generate a report to view resource usage
```
"runs, users" --> "runs. Users"

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+==========================
+
+Operations dashboard provides operational view for pipeline and CDAP application programs.
+CDAP does not offer a lot today in terms of a holistic view of all pipelines on a CDAP installation.
```
By today, you mean pre-5.0?

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+  - Reports summary also shows number of runs grouped by the namespace, app-type, start-type and owners.
+
+
+More information about using the Operations and Reporting application's usage can be found in the REST API docs.
```
insert a link to the relevant docs?

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+      - Easily monitor which pipelines have failed in the last 24 hours and at which time did the failures happen.
+
+  - Looking into the future pipeline resource usage
+      - Operations dashboard also provides a view of the programs that are scheduled to be run in the future,
```
why have a comma at the end of this line?

anwar6953 commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Example usage
+  - Look into pipelines that failed in the last week as a table
+  - Reports result contain two components, summary and program runs table.
+  - In authorization enabled clusters, administrators can generate report to summarize pipeline runs by users and
```
you mean authentication?
Only authentication is the requirement, and not authorization, right?

gsps1 replied:
its authorization enabled cluster as its principal name we group by.

anew replied:
so, it's impersonation, which requires authentication, but not authorization.

anwar6953 commented on cdap-docs/reference-manual/source/http-restful-api/index.rst:
```diff
@@ -84,6 +86,8 @@ data isolation. This is an inital step towards introducing `multi-tenancy
 - :doc:`Metrics: <metrics>` retrieving metrics for system and user applications (user-defined metrics)
 - :doc:`Monitor: <monitor>` checking the status of various system and custom CDAP services
+- :doc:`Operations Dashboard: <operations>` check in real time status of past program runs and future schedules to monitor cluster utilization
+- :doc:`Reporting: <reporting>` Generate reports to understand and monitor cluster usage, program runs and resource utilization
```
nit - inconsistent casing in "Generate". The previous lines don't case as such.

anwar6953 commented on cdap-docs/reference-manual/source/http-restful-api/index.rst:
```diff
@@ -105,8 +109,10 @@ data isolation. This is an inital step towards introducing `multi-tenancy
+- :doc:`Operations Dashboard: <operations>` checking the real time status of past program runs and future schedules to monitor cluster utilization
 - :doc:`Preferences: <preferences>` setting, retrieving, and deleting preferences
 - :doc:`Query: <query>` sending ad-hoc queries to CDAP datasets
+- :doc:`Reporting: <reporting>` Generate reports to understand and monitor cluster usage, program runs and resource utilization
```
nit - casing inconsistency in "Generate"

anwar6953 commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+   - Query start time in seconds
+ * - ``duration``
+   - Duration in seconds, query end time is inferred as startTime + duration in seconds
+ * - `namespace``
```
missing a ` before "namespace"

anwar6953 commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+return historical runs which were either started, running or reached end state (stopped/killed/failed) in that time window.
+
+For the future time range, time range after current time till the end time,
+operations dashboard would return all the programs that are scheduled to run in that time window,
```
after "time window", start a new sentence, instead of just using a comma.

anew commented on cdap-docs/admin-manual/source/operations/index.rst:
```diff
@@ -38,6 +39,10 @@ Operations
+.. |operations-dashboard| replace:: **Operations Dashboard and Reporting:**
+.. _operations-dashboard: operations-dashboard.html
+
+- |operations-dashboard|_ Operations with real-time dashboard and reporting view to look into program run statistics and cluster usage.
```
Remove "Operations with", and replace "view to look into" with "for". Also, can it be "resource usage"?

```
Real-time dashboard and reporting for program run statistics and resource usage.
```

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Operations Dashboard
+======================
+
+Operations dashboard provides users two views: Dashboard and Reporting.
```
delete this sentence

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+.. _operations-dashboard:
+
+======================
+Operations Dashboard
```
This section should be called "Dashboard and Reports", to reflect what it is in the UI

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+
+Operations dashboard provides users two views: Dashboard and Reporting.
+
+The Dashboard view is a real-time interactive interface that allows users to take a sneak peek
```
remove "view"

anew replied:
...that visualizes program run statistics and resource usage, over any 24-hour window in the last week or, for a more detailed view, over any 1-hour window in the last 24 hours.

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+of the program runs statistics, resource usage within any 24 hour time window in the last week.
+A user can also select any hour within the 24 hour view to get program run details within that hour.
+
+The Reporting view provides a comprehensive report of program runs. Users can generate a report to view resource usage
```
Reports provide comprehensive insights into program runs.

anew replied:
Also, "Dashboard" and "Reports" should be emphasized: ```**Dashboard**``` and ```**Reports**```.

anew replied:
I think it is better to address the reader in the second person. Instead of "users can generate", say "you can generate".

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+job will process the program runs and generate the report for the user.
+
+
+Operations Dashboard Usage
```
Just title this "Dashboard". This is the user guide so we don't need words like 'usage".

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Operations Dashboard Usage
+==========================
+
+Operations dashboard provides operational view for pipeline and CDAP application programs.
```
The Dashboard provides an operational view of CDAP pipelines and application programs.

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+==========================
+
+Operations dashboard provides operational view for pipeline and CDAP application programs.
+Before 5.0, CDAP does not offer a holistic view of all pipelines on a CDAP installation to monitor.
```
did not offer

anew replied:
Let's rephrase this. We do not need to motivate why it is here. Just say what it does:

```
The **Dashboard** was introduced in CDAP 5.0 to offer a holistic 
operational view of all pipelines and application programs running 
in a CDAP instance. It addresses these common operational needs:

- ...
```

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+CDAP Operations dashboard addresses those shortcomings and would be a useful tool to answer the following use cases and questions by administrators.
+
+  - CDAP System status
+      - Filtering and viewing pipelines that are currently running
```
either use infinitive or gerund form in all bullets, but don't mix them:

```
- Current Status
  - Filter and view pipelines...
  - Filter and view pipelines that...
  - Slice and dice ...
  - Monitor which ...
```

Remove the word "Easily", it adds no value here. (it would in a sales deck)

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+      - Slicing and dicing by time, program-status, namespace, user, start-method etc.
+      - Easily monitor which pipelines have failed in the last 24 hours and at which time did the failures happen.
+
+  - Looking into the future pipeline resource usage
```
Predicting future resource usage.

(this is not only for pipelines, right? any program if I understand correctly)

anew replied:
And to be consistent in style and remove boilerplate:

``` 
- Predict future resource usage
  - View program and pipeline runs scheduled in the future to predict and proactively manage resource usage.  
```

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+      - By peeking into future time range in dashboard, users can look into how many pipelines/programs will be run each hour, enabling them to predict and manage cluster usage efficiently.
+
+
+Reporting Application Usage
```
Reports

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Reporting Application Usage
+===========================
+
+Reporting offers tool for administrators to take a historical look at their cluster's resource usage, program runs and provides various
```
Reports are a tool for administrators to take a historical look at their applications' resource usage, program runs, and performance, optionally filtered by criteria such as program type or final run status.

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+filters that they could apply to gain insight into the pipelines and CDAP applications performance
+and resource utilization.
+
+Example usage
```
Reports contain two components: a summary and a program runs table. They provide, for example, this information:

- Tabular view of pipelines that failed in the last week
- In multi-tenant deployments, a summary and detailed report of pipelines run by each user
- Minimum, maximum and average duration of program runs that match the report request filters
- Summary of runs grouped by namespace, application type, start method and owner.

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+    look into the pipelines run by each user in the summary.
+  - Reports summary includes other useful information about the minimum, maximum and average duration
+    of program runs that match the report request filters.
+  - Reports summary also shows number of runs grouped by the namespace, app-type, start-type and owners.
```
what's the owner?

gsps1 replied:
its the user who ran the application, i agree owner is a wrong term to denote that. but currently its surfaced as owners in the UI, i have updated the doc to use the term "users".

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+
+
+More information about using the Operations can be found at :ref:`Operations REST API <http-restful-api-operations>`
+and Reporting application's usage can be found at :ref:`Reporting REST API <http-restful-api-reporting>` in the REST API docs.
```
Reporting/Reports x2

anew replied:
docs/documentation

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,60 @@
+Reporting offers tool for administrators to take a historical look at their cluster's resource usage, program runs and provides various
+filters that they could apply to gain insight into the pipelines and CDAP applications performance
+and resource utilization.
+
```
Don't you want to explain that Reports must be enabled before you can use them? And how to request a report? Isn't that an important part of this guide?

anew replied:
Is it enabled by namespace or for all namespaces? Are reports across namespaces or for a single namespace?

gsps1 replied:
its enabled only on the current namespace and can be configured to monitor more.  i have added additional docs to explain them.

anew reviewed (changes requested):
I have reviewed the guide, not the REST API so far. Is documentation for the REST API necessary? If so, please fix the wording and use what is used in the UI: "Reports" and "Dashboard" (not "Reporting" and "Operational Dashboard". Also see my other comments. 

Overall this reads more as a use case and requirements document than a user guide. Instead of listing what use cases it addresses, you should explain how to use it.

gsps1 commented:
@anew thanks for the review, i have addressed the comments and suggestions. please take a look again. about including the REST API for operations and reports, i added it to be consistent as we always document newly added API but here the use case is driven by UI, let me know if those can be removed.

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+**Dashboard** is a real-time interactive interface that visualizes program run statistics,
+over any 24-hour window in the last week or, for a more detailed view, over any 1-hour window.
+
+**Reports** provide comprehensive insights into program runs. you can generate a report to view
```
you->You

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+Dashboard
+=========
+
+The Dashboard provides an operational view of CDAP pipelines and application programs.
```
Remove this first sentence - it is repeated in the next sentence

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+Reports is a tool for administrators to take a historical look at their applications'
+program runs, and performance, optionally filtered by criteria such as program type or final run status.
+
+Reports must be enabled first, enabling reports, deploys the **cdap-program-report** application to system namespace
```
Reports must be enabled before you can use it. This deploys ...

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+program runs, and performance, optionally filtered by criteria such as program type or final run status.
+
+Reports must be enabled first, enabling reports, deploys the **cdap-program-report** application to system namespace
+and starts a Spark service, The spark service enables you to generate, list, view, save and delete reports.
```
Spark service. This service enables you...

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+
+Reports contain two components: a summary and a program runs table. They provide, for example, this information:
+
+  - Tabular view of program runs that matches your request filters, with selected column fields.
```
that match your request filters

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+  - Minimum, maximum and average duration of program runs that match the report request filters
+  - Summary of runs grouped by namespace, application type, start method and user.
+
+By default, reports expire 2 days after its creation time, unless they are explicitly saved.
```
after their creation

anew commented on cdap-docs/admin-manual/source/operations/operations-dashboard.rst:
```diff
@@ -0,0 +1,75 @@
+  - Reports is **not** supported on clusters with Spark version lower than 2.1.
+  - Only the program runs that were started in or after CDAP-5.0 will be processed and included in the reports, program runs from earlier CDAP version will be skipped.
+
+More information about using the Operations can be found at :ref:`Operations REST API <http-restful-api-operations>`
```
You means, about the Dashboard?

gsps1 replied:
yes, dashboard, i picked the incorrect word.

anew commented on cdap-docs/reference-manual/source/http-restful-api/index.rst:
```diff
@@ -84,6 +86,8 @@ data isolation. This is an inital step towards introducing `multi-tenancy
 - :doc:`Logging: <logging>` retrieving application logs
 - :doc:`Metrics: <metrics>` retrieving metrics for system and user applications (user-defined metrics)
 - :doc:`Monitor: <monitor>` checking the status of various system and custom CDAP services
+- :doc:`Operations Dashboard: <operations>` check in real time status of past program runs and predict future resource usage
```
Call it just "Dashboard" as the UI does.

anew commented on cdap-docs/reference-manual/source/http-restful-api/index.rst:
```diff
@@ -105,8 +109,10 @@ data isolation. This is an inital step towards introducing `multi-tenancy
 - :doc:`Metrics: <metrics>` retrieving metrics for system and user applications (user-defined metrics)
 - :doc:`Monitor: <monitor>` checking the status of various system and custom CDAP services
 - :doc:`Namespace: <namespace>` creating and managing namespaces
+- :doc:`Operations Dashboard: <operations>` check in real time status of past program runs and predict future resource usage
```
same here.

anew commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+.. include:: base-url.txt
+
+
+Operations HTTP REST API
```
Dashboard

anew commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+.. _http-restful-api-operations:
+
+=====================================
+Operations Dashboard HTTP RESTful API
```
Remove Operations

anew commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+ * - ``namespace``
+   - Set of namespaces, used for filtering by namespace, only programs that ran in this namespace or schedules that will be run in this namespace will be returned.
+
+For the given query timerange, for the time between start time and the current-time, operations dashboard would
```
change "operations dashboard would" to "the response will". And same in the next paragraph. 

Please never use "would", it is not very assertive and begs the question, it "would" under what conditions?

anew commented on cdap-docs/reference-manual/source/http-restful-api/operations.rst:
```diff
@@ -0,0 +1,96 @@
+return historical runs which were either started, running or reached end state (stopped/killed/failed) in that time window.
+
+For the future time range, time range after current time till the end time,
+operations dashboard would return all the programs that are scheduled to run in that time window.
```
Again: please change all occurrences of "operations dashboard" to "dashboard"

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+
+.. highlight:: console
+
+The Reports view provides administrators and developers a way to generate comprehensive report of program runs,
```
here you are documenting an API, not a view. So change it to: "Reports provide ..."

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+
+The Reports view provides administrators and developers a way to generate comprehensive report of program runs,
+specify filters based on various options such as namespaces, program status and type of program,
+within a given time range such as 24 hours or one week or a custom time range. Users can also specify the columns
```
**You** can also ... that **you** would like

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+within a given time range such as 24 hours or one week or a custom time range. Users can also specify the columns
+they would like to include in the generated report.
+
+The reports view is backed by ReportGenerationApp system application.
```
**Reports are** backed by **the** ReportGenerationApp system application **, which is a Spark service (named ReportGenerationSpark) with** endpoints to generate, ....

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+   ]
+ }
+
+On executing this generate reports endpoint, backend asynchronously launches spark job to process and generate report and immediately responds with report-id JSON.
```
**On submitting this request, the** backend ... launches **a Spark** job ... generate **the** report. **The immediate response contains the report id.**

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+    "id": <report-id>
+ }
+
+Users can use this report-id to check on status of the report, if it is running, completed or failed using info endpoint
```
**You** can ... check **the** status ... failed **, using the** ```info``` endpoint

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+---------------
+
+On Authentication enabled clusters, reports are isolated by logged-in users who generated the reports,
+If the user wishes to share a generated report, they can get an encrypted
```
**If you wish** ..., **you can generate** an ... share-id **that can be used by others to access the report.**

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+Viewing Report Status
+----------------------
+
+The info endpoint can be used to get the report status and summary for a report,
```
report. (period)

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+----------------------
+
+The info endpoint can be used to get the report status and summary for a report,
+report info can be queried either by providing report-id directly or by providing share-id through query param.
```
**It** can ... providing **the** report-id or ... **a** share-id ....

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+   "status":<report-status>
+ }
+
+For currently active reports, the response would not have report summary
```
... the response **does not contain a** report summary.

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+   "status":"RUNNING"
+ }
+
+For failed reports, the response would have error information.
```
.. the response **contains the** error information.

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+Downloading Report contents
+----------------------------
+
+Download endpoint can be used to download the report contents,
```
The Download endpoint ... contents. (period).

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+----------------------------
+
+Download endpoint can be used to download the report contents,
+report can be downloaded either by providing report-id directly or by providing share-id through query param.
```
You can either provide a report id or a share id through a query parameter.

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+
+Saving Reports
+--------------
+By default reports are expired after 2 weeks, however they can be saved by using the following REST endpoint,
```
are expired/expire

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+
+::
+
+ DELETE : /v3/namespaces/system/apps/ReportGenerationApp/spark/ReportGenerationSpark/methods/reports/report-id/
```
should there be no trailing / ?

anew commented on cdap-docs/reference-manual/source/http-restful-api/reporting.rst:
```diff
@@ -0,0 +1,440 @@
+
+Deleting Reports
+----------------
+If reports either generated or failed, deemed unnecessary, they can be deleted by making a call to the following endpoint
```
Unneeded reports (whether generated or failed) can be deleted...

anew reviewed (changes requested):
Please see my comments
